File: core/config/workspaces.go
// Package config provides workspace management for elmos.
// Workspaces are isolated directories for building specific boards.
package config

import (
	"fmt"
	"os"
	"path/filepath"
)

// WorkspaceManager handles workspace-related operations.
type WorkspaceManager struct {
	rootPath string // Path to workspace root (where .elmos/ exists)
}

// NewWorkspaceManager creates a new WorkspaceManager.
// rootPath should point to the workspace root directory.
func NewWorkspaceManager(rootPath string) *WorkspaceManager {
	if rootPath == "" {
		rootPath = "."
	}
	return &WorkspaceManager{
		rootPath: rootPath,
	}
}

// FindWorkspaceRoot searches for .elmos/ directory starting from cwd and going up.
// Returns the workspace root path or error if not found.
func FindWorkspaceRoot() (string, error) {
	cwd, err := os.Getwd()
	if err != nil {
		return "", err
	}

	// Search up the directory tree for .elmos/
	for {
		elmosPath := filepath.Join(cwd, ".elmos")
		if info, err := os.Stat(elmosPath); err == nil && info.IsDir() {
			return cwd, nil
		}

		parent := filepath.Dir(cwd)
		if parent == cwd {
			// Reached root directory
			break
		}
		cwd = parent
	}

	return "", fmt.Errorf("workspace root (.elmos/) not found in current directory or any parent")
}

// GetRoot returns the workspace root path.
func (wm *WorkspaceManager) GetRoot() string {
	return wm.rootPath
}

// GetElmosDir returns the .elmos/ directory path.
func (wm *WorkspaceManager) GetElmosDir() string {
	return filepath.Join(wm.rootPath, ".elmos")
}

// GetConfigDir returns the .elmos/config directory path.
func (wm *WorkspaceManager) GetConfigDir() string {
	return filepath.Join(wm.GetElmosDir(), "config")
}

// GetCacheDir returns the .elmos/build-cache directory path.
func (wm *WorkspaceManager) GetCacheDir() string {
	return filepath.Join(wm.GetElmosDir(), "build-cache")
}

// GetMachineDir returns the directory for machine definitions.
func (wm *WorkspaceManager) GetMachineDir() string {
	return filepath.Join(wm.rootPath, "machine")
}

// GetMachineDefPath returns the path to a specific machine definition file.
func (wm *WorkspaceManager) GetMachineDefPath(machineID string) string {
	return filepath.Join(wm.GetMachineDir(), fmt.Sprintf("%s.yml", machineID))
}

// GetPluginsDir returns the directory for user plugins.
func (wm *WorkspaceManager) GetPluginsDir() string {
	return filepath.Join(wm.rootPath, "plugins")
}

// GetDefaultConfigPath returns the path to default.yml.
func (wm *WorkspaceManager) GetDefaultConfigPath() string {
	return filepath.Join(wm.rootPath, "default.yml")
}

// GetPluginConfigPath returns the path to .elmos/plugins.yml.
func (wm *WorkspaceManager) GetPluginConfigPath() string {
	return filepath.Join(wm.GetElmosDir(), "plugins.yml")
}

// Initialize creates the workspace directory structure.
// Creates .elmos/, machine/, plugins/ directories if they don't exist.
func (wm *WorkspaceManager) Initialize() error {
	dirs := []string{
		wm.GetElmosDir(),
		wm.GetConfigDir(),
		wm.GetCacheDir(),
		wm.GetMachineDir(),
		wm.GetPluginsDir(),
	}

	for _, dir := range dirs {
		if err := os.MkdirAll(dir, 0755); err != nil {
			return fmt.Errorf("failed to create directory %s: %w", dir, err)
		}
	}

	return nil
}

// Exists checks if the workspace is initialized (has .elmos/ directory).
func (wm *WorkspaceManager) Exists() bool {
	elmosDir := wm.GetElmosDir()
	info, err := os.Stat(elmosDir)
	return err == nil && info.IsDir()
}

// EnsureInitialized verifies the workspace is properly initialized.
// Returns error if critical directories are missing.
func (wm *WorkspaceManager) EnsureInitialized() error {
	if !wm.Exists() {
		return fmt.Errorf("workspace not initialized at %s", wm.rootPath)
	}

	// Check for critical subdirs
	requiredDirs := []string{
		wm.GetElmosDir(),
		wm.GetCacheDir(),
	}

	for _, dir := range requiredDirs {
		if info, err := os.Stat(dir); err != nil || !info.IsDir() {
			return fmt.Errorf("workspace directory missing or invalid: %s", dir)
		}
	}

	return nil
}

// CleanCache deletes the build cache directory.
// Useful for forcing a clean rebuild.
func (wm *WorkspaceManager) CleanCache() error {
	cacheDir := wm.GetCacheDir()
	if err := os.RemoveAll(cacheDir); err != nil {
		return fmt.Errorf("failed to clean cache: %w", err)
	}

	// Recreate the empty cache directory
	return os.MkdirAll(cacheDir, 0755)
}

// GetCachePath returns the path for a specific cache artifact.
func (wm *WorkspaceManager) GetCachePath(fingerprint string) string {
	return filepath.Join(wm.GetCacheDir(), fingerprint)
}

// GetCacheIndexPath returns the path to the build cache index file.
func (wm *WorkspaceManager) GetCacheIndexPath() string {
	return filepath.Join(wm.GetCacheDir(), "index.json")
}

// QueryMachines returns all machine definitions in the workspace.
func (wm *WorkspaceManager) QueryMachines() ([]string, error) {
	machineDir := wm.GetMachineDir()

	// Return empty list if machine directory doesn't exist yet
	if _, err := os.Stat(machineDir); err != nil {
		return []string{}, nil
	}

	entries, err := os.ReadDir(machineDir)
	if err != nil {
		return nil, fmt.Errorf("failed to read machine directory: %w", err)
	}

	var machines []string
	for _, entry := range entries {
		if !entry.IsDir() && filepath.Ext(entry.Name()) == ".yml" {
			name := entry.Name()[:len(entry.Name())-4] // Remove .yml
			machines = append(machines, name)
		}
	}

	return machines, nil
}

File: src/config/workspaces.cpp
// ============================================================================
// config/workspaces.cpp — Workspace management implementation
// ============================================================================

#include "workspaces.hpp"

#include <algorithm>
#include <cstdlib>
#include <fstream>

#if defined(ELMOS_PLATFORM_DARWIN) || defined(ELMOS_PLATFORM_LINUX)
#include <sys/types.h>
#include <unistd.h>

#include <pwd.h>
#endif

namespace elmos::config {

namespace fs = std::filesystem;

namespace {

auto resolve_registry_home() -> std::string {
#if defined(ELMOS_PLATFORM_DARWIN) || defined(ELMOS_PLATFORM_LINUX)
    // When running under sudo, keep workspace registry tied to the invoking user.
    if (const char* sudo_user = std::getenv("SUDO_USER"); sudo_user && *sudo_user) {
        if (auto* pw = ::getpwnam(sudo_user); pw && pw->pw_dir && *pw->pw_dir) {
            return pw->pw_dir;
        }
    }

    if (auto* pw = ::getpwuid(::getuid()); pw && pw->pw_dir && *pw->pw_dir) {
        return pw->pw_dir;
    }
#endif

    if (const char* home = std::getenv("HOME"); home && *home) {
        return home;
    }

    return "/tmp";
}

}  // namespace

WorkspaceManager::WorkspaceManager(const std::string& root_path)
    : root_path_(root_path.empty() ? "." : root_path) {}

auto WorkspaceManager::get_root() const -> std::string {
    return root_path_;
}
auto WorkspaceManager::get_elmos_dir() const -> std::string {
    return (fs::path(root_path_) / ".elmos").string();
}
auto WorkspaceManager::get_config_dir() const -> std::string {
    return (fs::path(get_elmos_dir()) / "config").string();
}
auto WorkspaceManager::get_cache_dir() const -> std::string {
    return (fs::path(get_elmos_dir()) / "build-cache").string();
}
auto WorkspaceManager::get_machine_dir() const -> std::string {
    return (fs::path(root_path_) / "machine").string();
}
auto WorkspaceManager::get_machine_def_path(const std::string& machine_id) const -> std::string {
    return (fs::path(get_machine_dir()) / (machine_id + ".yml")).string();
}
auto WorkspaceManager::get_plugins_dir() const -> std::string {
    return (fs::path(root_path_) / "plugins").string();
}
auto WorkspaceManager::get_default_config_path() const -> std::string {
    return (fs::path(root_path_) / "default.yml").string();
}
auto WorkspaceManager::get_plugin_config_path() const -> std::string {
    return (fs::path(get_elmos_dir()) / "plugins.yml").string();
}
auto WorkspaceManager::get_cache_path(const std::string& fingerprint) const -> std::string {
    return (fs::path(get_cache_dir()) / fingerprint).string();
}
auto WorkspaceManager::get_cache_index_path() const -> std::string {
    return (fs::path(get_cache_dir()) / "index.json").string();
}

auto WorkspaceManager::initialize() -> VoidResult {
    std::error_code ec;
    for (const auto& dir : {get_elmos_dir(), get_config_dir(), get_cache_dir(), get_machine_dir(),
                            get_plugins_dir()}) {
        fs::create_directories(dir, ec);
        if (ec) {
            return make_error(
                Error::config("failed to create directory " + dir + ": " + ec.message()));
        }
    }
    return {};
}

auto WorkspaceManager::exists() const -> bool {
    std::error_code ec;
    return fs::is_directory(get_elmos_dir(), ec);
}

auto WorkspaceManager::ensure_initialized() -> VoidResult {
    if (!exists()) {
        return make_error(Error::config("workspace not initialized at " + root_path_));
    }
    for (const auto& dir : {get_elmos_dir(), get_cache_dir()}) {
        std::error_code ec;
        if (!fs::is_directory(dir, ec)) {
            return make_error(Error::config("workspace directory missing: " + dir));
        }
    }
    return {};
}

auto WorkspaceManager::clean_cache() -> VoidResult {
    std::error_code ec;
    fs::remove_all(get_cache_dir(), ec);
    if (ec) {
        return make_error(Error::config("failed to clean cache: " + ec.message()));
    }
    fs::create_directories(get_cache_dir(), ec);
    return {};
}

auto WorkspaceManager::query_machines() -> Result<std::vector<std::string>> {
    auto machine_dir = get_machine_dir();
    std::error_code ec;
    if (!fs::exists(machine_dir, ec))
        return std::vector<std::string>{};

    std::vector<std::string> machines;
    for (const auto& entry : fs::directory_iterator(machine_dir, ec)) {
        if (!entry.is_directory() && entry.path().extension() == ".yml") {
            machines.push_back(entry.path().stem().string());
        }
    }
    if (ec) {
        return make_error(Error::config("failed to read machine directory: " + ec.message()));
    }
    return machines;
}

auto WorkspaceManager::find_workspace_root() -> Result<std::string> {
    std::error_code ec;
    auto cwd = fs::current_path(ec);
    if (ec) {
        return make_error(Error::config("cannot get working directory: " + ec.message()));
    }

    auto current = cwd;
    while (true) {
        if (fs::is_directory(current / ".elmos", ec)) {
            return current.string();
        }
        auto parent = current.parent_path();
        if (parent == current)
            break;
        current = parent;
    }

    return make_error(
        Error::config("workspace root (.elmos/) not found in current directory or any parent"));
}

// ── Global workspace registry ───────────────────────────────────────────────

auto WorkspaceManager::global_elmos_dir() -> std::string {
    return resolve_registry_home() + "/.elmos";
}

auto WorkspaceManager::workspace_config_path(const std::string& name) -> std::string {
    return global_elmos_dir() + "/workspaces/" + name + "/config.yaml";
}

auto WorkspaceManager::get_active_workspace() -> Result<std::string> {
    auto path = global_elmos_dir() + "/active";
    std::ifstream f(path);
    if (!f)
        return make_error(Error::config("no active workspace — run 'elmos init <name>' first"));
    std::string name;
    std::getline(f, name);
    while (!name.empty() && (name.back() == '\n' || name.back() == '\r' || name.back() == ' '))
        name.pop_back();
    if (name.empty())
        return make_error(Error::config("no active workspace — run 'elmos init <name>' first"));
    return name;
}

auto WorkspaceManager::set_active_workspace(const std::string& name) -> VoidResult {
    auto dir = global_elmos_dir() + "/workspaces";
    std::error_code ec;
    fs::create_directories(dir, ec);

    auto cfg_path = workspace_config_path(name);
    if (!fs::exists(cfg_path)) {
        return make_error(
            Error::config("workspace '" + name + "' not found — run 'elmos init " + name + "'"));
    }

    auto path = global_elmos_dir() + "/active";
    std::ofstream f(path, std::ios::trunc);
    if (!f)
        return make_error(Error::config("cannot write " + path));
    f << name;
    return {};
}

auto WorkspaceManager::workspace_dir(const std::string& name) -> std::string {
    return global_elmos_dir() + "/workspaces/" + name;
}

auto WorkspaceManager::list_workspaces() -> Result<std::vector<std::string>> {
    auto dir = global_elmos_dir() + "/workspaces";
    std::error_code ec;
    if (!fs::is_directory(dir, ec))
        return std::vector<std::string>{};

    std::vector<std::string> names;
    for (const auto& entry : fs::directory_iterator(dir, ec)) {
        // Each workspace is a subdirectory containing config.yaml
        if (entry.is_directory() && fs::exists(entry.path() / "config.yaml", ec))
            names.push_back(entry.path().filename().string());
    }
    std::sort(names.begin(), names.end());
    return names;
}

}  // namespace elmos::config
